streams/src/main/java/org/apache/kafka/streams/state/internals/WindowKeySchema.java

```diff
@@ -93,4 +97,107 @@ public boolean hasNext(final KeyValueIterator<Bytes, ?> iterator) {
+        System.arraycopy(binaryKey, 0, bytes, 0, bytes.length);
+        final K key = deserializer.deserialize(topic, bytes);
+        final Window window = extractWindow(binaryKey, windowSize);
+        return new Windowed<>(key, window);
+    }
+
+    private static Window extractWindow(final byte[] binaryKey, final long windowSize) {
+        final ByteBuffer buffer = ByteBuffer.wrap(binaryKey);
+        final long start = buffer.getLong(binaryKey.length - TIMESTAMP_SIZE);
+        return timeWindowForSize(start, windowSize);
+    }
+
+    // for store serdes
+
+    public static Bytes toStoreKeyBinary(final Bytes key, final long timestamp, final int seqnum) {
+        byte[] serializedKey = key.get();
+        return toStoreKeyBinary(serializedKey, timestamp, seqnum);
+    }
+
+    public static <K> Bytes toStoreKeyBinary(final K key, final long timestamp, final int seqnum, final StateSerdes<K, ?> serdes) {
+        byte[] serializedKey = serdes.rawKey(key);
+        return toStoreKeyBinary(serializedKey, timestamp, seqnum);
+    }
+
+    public static Bytes toStoreKeyBinary(final Windowed<Bytes> timeKey, final int seqnum) {
+        final byte[] bytes = timeKey.key().get();
+        return toStoreKeyBinary(bytes, timeKey.window().start(), seqnum);
+    }
+
+    public static <K> Bytes toStoreKeyBinary(final Windowed<K> timeKey, final int seqnum, final StateSerdes<K, ?> serdes) {
+        byte[] serializedKey = serdes.rawKey(timeKey.key());
+        return toStoreKeyBinary(serializedKey, timeKey.window().start(), seqnum);
+    }
+
+    // package private for testing
+    static Bytes toStoreKeyBinary(byte[] serializedKey, final long timestamp, final int seqnum) {
+        ByteBuffer buf = ByteBuffer.allocate(serializedKey.length + TIMESTAMP_SIZE + SEQNUM_SIZE);
+        buf.put(serializedKey);
+        buf.putLong(timestamp);
+        buf.putInt(seqnum);
+
+        return Bytes.wrap(buf.array());
+    }
+
+    public static byte[] extractStoreKeyBytes(final byte[] binaryKey) {
+        final byte[] bytes = new byte[binaryKey.length - TIMESTAMP_SIZE - SEQNUM_SIZE];
+        System.arraycopy(binaryKey, 0, bytes, 0, bytes.length);
+        return bytes;
+    }
+
+    public static <K> K extractStoreKey(final byte[] binaryKey, final StateSerdes<K, ?> serdes) {
+        final byte[] bytes = new byte[binaryKey.length - TIMESTAMP_SIZE - SEQNUM_SIZE];
+        System.arraycopy(binaryKey, 0, bytes, 0, bytes.length);
+        return serdes.keyFrom(bytes);
+    }
+
+    public static long extractStoreTimestamp(byte[] binaryKey) {
+        return ByteBuffer.wrap(binaryKey).getLong(binaryKey.length - TIMESTAMP_SIZE - SEQNUM_SIZE);
+    }
+
+    public static int extractStoreSequence(byte[] binaryKey) {
+        return ByteBuffer.wrap(binaryKey).getInt(binaryKey.length - SEQNUM_SIZE);
+    }
+
+    public static <K> Windowed<K> fromStoreKey(final byte[] binaryKey, final long windowSize, final StateSerdes<K, ?> serdes) {
```

Comment:
nit: parameters on separate lines

Reply:
As for parameters: I think for this it does not matter much since most of them are less than three and a single line can fit.

Reply:
I would prefer multiple lines, too -- it makes the diff easier to read if parameters are added/removed